src/aranges.rs

```diff
@@ -0,0 +1,249 @@
+    ///
+    /// # let buf = [];
+    /// # let read_debug_aranges_section_somehow = || &buf;
+    /// let debug_aranges = DebugAranges::<LittleEndian>::new(read_debug_aranges_section_somehow());
+    ///
+    /// for parse_result in debug_aranges.aranges() {
+    ///   let arange = parse_result.unwrap();
+    ///   println!("arange starts at {}, has length {}", arange.start(), arange.len());
+    /// }
+    /// ```
+    pub fn aranges(&self) -> ArangesIter<Endian> {
+        ArangesIter { input: self.debug_aranges_section }
+    }
+}
+
+pub struct ArangesIter<'input, Endian>
+    where Endian: Endianity
+{
+    input: EndianBuf<'input, Endian>
+}
+
+impl<'input, Endian> Iterator for ArangesIter<'input, Endian>
+    where Endian: Endianity
+{
+    type Item = ParseResult<Vec<Arange>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.input.is_empty() {
+            None
+        } else {
+            match Arange::parse(self.input) {
+                Ok((remaining_input, aranges)) => {
+                    self.input = remaining_input;
+                    Some(Ok(aranges))
+                }
+                Err(e) => {
+                    self.input = self.input.range_to(..0);
+                    Some(Err(e))
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+struct ArangeHeader
+{
+    format: Format,
+    length: u64,
+    version: u16,
+    offset: DebugInfoOffset,
+    address_size: u8,
+    segment_size: u8,
+}
+
+#[derive(Debug, Clone, Eq)]
+pub struct Arange
+{
+    segment: u64,
+    offset: u64,
+    length: u64,
+    header: Rc<ArangeHeader>,
+}
+
+impl Arange
+{
+    /// Parse an arange set header. Returns a tuple of the remaining arange sets, the aranges to be
+    /// parsed for this set, and the newly created ArangeHeader struct.
+    fn parse_header<Endian>(input: EndianBuf<Endian>)
+                            -> ParseResult<(EndianBuf<Endian>,
+                                            EndianBuf<Endian>,
+                                            Rc<ArangeHeader>)>
+        where Endian: Endianity
+    {
+        let (rest, (length, format)) = try!(parse_unit_length(input.into()));
+        let (rest, version) = try!(parse_u16(rest.into()));
+
+        if version != 2 {
+            return Err(Error::UnknownVersion)
+        }
+
+        let (rest, offset) = try!(parse_debug_info_offset(rest.into(), format));
+        let (rest, address_size) = try!(parse_address_size(rest.into()));
+        let (rest, segment_size) = try!(parse_address_size(rest.into()));
+
+        let header_length = match format {
+            Format::Dwarf32 => 8,
+            Format::Dwarf64 => 12,
+        };
+        let dividing_line: usize = try!(length.checked_sub(header_length).ok_or(Error::BadLength)) as usize;
+
+        Ok((rest.range_from(dividing_line..),
+            rest.range_to(..dividing_line),
+            Rc::new(ArangeHeader { format: format,
```

Comment:
The reason I didn't do something like this originally is because there's nothing in the final parsed structs (`ArangeHeader`s and `Arange`s) that require keeping alive the original input buffer once parsing is complete.  Whereas making this change would require that.  Did you consider this tradeoff?